### PR TITLE
Enable PHPStan in tests

### PR DESCRIPTION
**Pull request by Majkl578**
|      Q       |   A
|------------- | -----------
| Type         | improvement
| BC Break     | no

#### Summary

Enabling PHPStan in tests/ at level 3, just like for lib/. :tada: 
Rewritten some Prophecy-based unit tests into PHPUnit mock system to avoid having another (unofficial) extension for Prophecy.

Replaces incomplete #3203 with no changes like (_Class_ -> _mixed_).

Currently depends on my fork of phpstan-phpunit due to https://github.com/phpstan/phpstan-phpunit/pull/28.

### CODE REVIEW

File: tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php [hunk 1]
```diff
@@ -743,11 +743,11 @@ public function testFetchAllSupportFetchClass()
 
         $results = $stmt->fetchAll(
             FetchMode::CUSTOM_OBJECT,
-            __NAMESPACE__.'\\MyFetchClass'
+            MyFetchClass::class
         );
 
         self::assertCount(1, $results);
-        self::assertInstanceOf(__NAMESPACE__.'\\MyFetchClass', $results[0]);
+        self::assertInstanceOf(MyFetchClass::class, $results[0]);
```
Comment (Majkl578) [inline tests/Doctrine/Tests/DBAL/Functional/DataAccessTest.php]: Not strictly needed, but affected by https://github.com/phpstan/phpstan-phpunit/issues/27.

File: composer.json [hunk 2]
```diff
@@ -21,6 +21,7 @@
         "doctrine/coding-standard": "^4.0",
         "jetbrains/phpstorm-stubs": "^2018.1.2",
         "phpstan/phpstan": "^0.10.1",
+        "phpstan/phpstan-phpunit": "dev-GetMockBuilderDynamicReturnTypeExtension-with-undefined-class as 0.10.0.999",
```
Comment (morozov) [inline composer.json]: Note to self: make sure this patch is released first.

File: tests/Doctrine/Tests/DBAL/Driver/PDOExceptionTest.php [hunk 3]
```diff
@@ -24,7 +24,7 @@ class PDOExceptionTest extends DbalTestCase
     /**
      * The wrapped PDO exception mock.
      *
-     * @var \PDOException|\PHPUnit_Framework_MockObject_MockObject
+     * @var \PDOException
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Driver/PDOExceptionTest.php]: Yay!

File: tests/Doctrine/Tests/DBAL/Functional/Driver/IBMDB2/DB2StatementTest.php [hunk 4]
```diff
@@ -27,6 +29,7 @@ protected function setUp()
     public function testExecutionErrorsAreNotSuppressed()
     {
         $stmt = $this->_conn->prepare('SELECT * FROM SYSIBM.SYSDUMMY1 WHERE \'foo\' = ?');
+        assert($stmt instanceof Statement);
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Functional/Driver/IBMDB2/DB2StatementTest.php]: What else could it be?

File: tests/Doctrine/Tests/DBAL/Schema/MySqlSchemaManagerTest.php [hunk 5]
```diff
@@ -15,7 +16,7 @@ class MySqlSchemaManagerTest extends \PHPUnit\Framework\TestCase
      */
     private $manager;
 
-    /** @var Connection */
+    /** @var Connection|MockObject */
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Schema/MySqlSchemaManagerTest.php]: It looks like the opposite of https://github.com/doctrine/dbal/pull/3260/files#diff-90748caec80a98aa0fcfef640916661bL27.

File: tests/Doctrine/Tests/DBAL/Schema/SchemaTest.php [hunk 6]
```diff
@@ -449,6 +451,9 @@ public function testVisitsNamespaceVisitor()
         $visitor->expects($this->exactly(2))
             ->method('acceptSequence');
 
-        self::assertNull($schema->visit($visitor));
+        $schema->visit($visitor);
+
+        self::doesNotPerformAssertions(); // FIXME
+
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Schema/SchemaTest.php]: Redundant empty line.

File: tests/Doctrine/Tests/DBAL/Types/DateImmutableTypeTest.php [hunk 7]
```diff
@@ -24,7 +25,7 @@ class DateImmutableTypeTest extends \PHPUnit\Framework\TestCase
     protected function setUp()
     {
         $this->type = Type::getType('date_immutable');
-        $this->platform = $this->prophesize(AbstractPlatform::class);
+        $this->platform = $this->getMockBuilder(AbstractPlatform::class)->getMock();
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Types/DateImmutableTypeTest.php]: Probably could use `$this->createMock()` instead.

File: tests/Doctrine/Tests/Mocks/TaskMock.php [hunk 8]
```diff
@@ -1,84 +0,0 @@
...
-
-namespace Doctrine\Tests\Mocks;
-
-use Doctrine\Common\Cli\AbstractNamespace;
-
-/**
- * TaskMock used for testing the CLI interface.
- * @author Nils Adermann <naderman@naderman.de>
- */
-class TaskMock extends \Doctrine\Common\Cli\Tasks\AbstractTask
```
Comment (morozov) [inline tests/Doctrine/Tests/Mocks/TaskMock.php]: :heart:

File: tests/Doctrine/Tests/DBAL/Types/DateImmutableTypeTest.php (hunk 7 above)
Comment (Majkl578) [reply tests/Doctrine/Tests/DBAL/Types/DateImmutableTypeTest.php]: Yes, forgot to change this file. :smiley:

File: tests/Doctrine/Tests/DBAL/Schema/MySqlSchemaManagerTest.php (hunk 5 above)
Comment (morozov) [reply tests/Doctrine/Tests/DBAL/Schema/MySqlSchemaManagerTest.php]: I see. Makes sense.

I thought, using `phpstan-phpunit` would eliminate the necessity to explicitly annotate mocks as such. Looks like it's only the case for the return values of methods like `createMock()` which automatically get the traits of the object under test and the mock. But for properties, you still have to specify the type explictly. Does it sound about right?

Comment (ondrejmirtes) [reply tests/Doctrine/Tests/DBAL/Schema/MySqlSchemaManagerTest.php]: Yes, it converts them in this case, currently hardcoded in PHPStan, but will be cleaner and part of the phpunit extension in 0.11.

**Comment by Majkl578** [comment]
Rebased, green on level 3, 65 more errors on level 7.

**Comment by morozov** [comment]
> Rebased on top of develop, green on level 3, 65 more errors on level 7.

We can wrap it up at level 3 and continue to improve later in order to avoid future conflicts and new issues on Level 3 and lower. BTW, why `develop`? Is it possible to move it to `master`, same as `lib/`, especially given that we should have more freedom in making changes in tests?

**Comment by Majkl578** [comment]
> BTW, why develop?

Picked wrong target when rebasing and didn't want to start again on top of master. :|

**Comment by morozov** [comment]
> Picked wrong target when rebasing and didn't want to start again on top of master. :|

Could you at least roll it back? I would like to have it done in `master`, and backporting to `master` and rebasing `develop` on top of it would be unnecessary extra work.

**Comment by morozov** [comment]
Closing in favor of #3582. Thanks, @Majkl578.

**Pull request closed without merging**